python/dgl/homophily.py

```diff
@@ -52,6 +58,139 @@ def node_homophily(graph, y):
+            fn.copy_e("same_class", "m"), fn.mean("m", "same_class_deg")
         )
-        return graph.ndata["node_value"].mean().item()
+        return F.as_scalar(F.mean(graph.ndata["same_class_deg"], dim=0))
+
+
+def edge_homophily(graph, y):
+    r"""Homophily measure from `Beyond Homophily in Graph Neural Networks:
+    Current Limitations and Effective Designs
+    <https://arxiv.org/abs/2006.11468>`__
+
+    Mathematically it is defined as follows:
+
+    .. math::
+      \frac{| \{ (u,v) : (u,v) \in \mathcal{E} \wedge y_u = y_v \} | }
+      {|\mathcal{E}|}
+
+    where :math:`\mathcal{E}` is the set of edges, and :math:`y_u` is the class
+    of node :math:`u`.
+
+    Parameters
+    ----------
+    graph : DGLGraph
+        The graph
+    y : Tensor
+        The node labels, which is a tensor of shape (|V|)
+
+    Returns
+    -------
+    float
+        The edge homophily ratio value
+
+    Examples
+    --------
+    >>> import dgl
+    >>> import torch
+
+    >>> graph = dgl.graph(([1, 2, 0, 4], [0, 1, 2, 3]))
+    >>> y = torch.tensor([0, 0, 0, 0, 1])
+    >>> dgl.edge_homophily(graph, y)
+    0.75
+    """
+    with graph.local_scope():
+        src, dst = graph.edges()
+        # Handle the case where graph is of dtype int32.
+        src = F.astype(src, F.int64)
+        dst = F.astype(dst, F.int64)
+        # Compute y_v = y_u for all edges.
+        edge_indicator = F.astype(y[src] == y[dst], F.float32)
+        return F.as_scalar(F.mean(edge_indicator, dim=0))
+
+
+def linkx_homophily(graph, y):
+    r"""Homophily measure from `Large Scale Learning on Non-Homophilous Graphs:
+    New Benchmarks and Strong Simple Methods
+    <https://arxiv.org/abs/2110.14446>`__
+
+    Mathematically it is defined as follows:
+
+    .. math::
+      \frac{1}{C-1} \sum_{k=1}^{C} \max \left(0, \frac{\sum_{v\in C_k}|\{u\in
+      \mathcal{N}(v): y_v = y_u \}|}{\sum_{v\in C_k}|\mathcal{N}(v)|} -
+      \frac{|\mathcal{C}_k|}{|\mathcal{V}|} \right)
+
+    where :math:`C` is the number of node classes, :math:`C_k` is the set of
+    nodes that belong to class k, :math:`\mathcal{N}(v)` is the predecessors of
+    node :math:`v`, :math:`y_v` is the class of node :math:`v`, and
+    :math:`\mathcal{V}` is the set of nodes.
+
+    Parameters
+    ----------
+    graph : DGLGraph
+        The graph
+    y : Tensor
+        The node labels, which is a tensor of shape (|V|)
+
+    Returns
+    -------
+    float
+        The homophily value
+
+    Examples
+    --------
+    >>> import dgl
+    >>> import torch
+
+    >>> graph = dgl.graph(([0, 1, 2, 3], [1, 2, 0, 4]))
+    >>> y = torch.tensor([0, 0, 0, 0, 1])
+    >>> dgl.linkx_homophily(graph, y)
+    0.19999998807907104
+    """
+    with graph.local_scope():
+        # Compute |{u\in N(v): y_v = y_u}| for each node v.
+        src, dst = graph.edges()
+        # Handle the case where graph is of dtype int32.
+        src = src.long()
```

Comment:
why not?
```
src = F.astype(src, F.int64)
```
I guess they are equivalent on PyTorch?

Create a local method? 
```
def edges_astype(graph)
        src, dst = graph.edges()
        # Handle the case where graph is of dtype int32.
        src = F.astype(src, F.int64)
        dst = F.astype(dst, F.int64)
        return src, dst
```

Or better name?

Reply:
For the first question, you are right. The reason for the alternative is that this method will only work for PyTorch and therefore I turned to more PyTorch-native syntax in case users want to check and understand the implementation.

Reply:
I've addressed the second question.